File: frontend/src/api/factoryRoutingApi.ts
import { apiClient } from '@/api/client'
import type { ApiResponse } from '@/api/authApi'

export interface FactoryRoutingResponse {
  routingId: number
  factoryName: string
  lineName: string
  operationSeq: number
  operationName: string
  routingStatus: FactoryRoutingStatus
  createdAt: string
  lastExecutionAt: string | null
}

export type FactoryRoutingStatus = 'ACTIVE' | 'INACTIVE'

export interface FactoryRoutingRequest {
  factoryName: string
  lineName: string
  operationSeq: number
  operationName: string
}

export interface FactoryRoutingTreeResponse {
  factoryName: string
  lines: FactoryRoutingLineResponse[]
}

export interface FactoryRoutingLineResponse {
  lineName: string
  operations: FactoryRoutingOperationResponse[]
}

export interface FactoryRoutingOperationResponse {
  routingId: number
  operationSeq: number
  operationName: string
}

export interface FactoryRoutingSearchParams {
  factoryName?: string
  lineName?: string
  routingStatus?: FactoryRoutingStatus
}

export interface FactoryRoutingStatusUpdateRequest {
  routingStatus: FactoryRoutingStatus
}

export interface FactoryRoutingUsageResponse {
  routingId: number
  workOrderCount: number
  executionCount: number
  workOrderNos: string[]
  executionIds: number[]
  canUpdate: boolean
  canDelete: boolean
  recommendedAction: string
}

function encodePath(value: string) {
  return encodeURIComponent(value)
}

export const factoryRoutingApi = {
  async getRoutings(params: FactoryRoutingSearchParams = {}) {
    const response = await apiClient.get<ApiResponse<FactoryRoutingResponse[]>>('/api/routings', {
      params
    })
    return response.data
  },

  async getRouting(id: number) {
    const response = await apiClient.get<ApiResponse<FactoryRoutingResponse>>(`/api/routings/${id}`)
    return response.data
  },

  async createRouting(request: FactoryRoutingRequest) {
    const response = await apiClient.post<ApiResponse<FactoryRoutingResponse>>('/api/routings', request)
    return response.data
  },

  async updateRouting(id: number, request: FactoryRoutingRequest) {
    const response = await apiClient.put<ApiResponse<FactoryRoutingResponse>>(`/api/routings/${id}`, request)
    return response.data
  },

  async deleteRouting(id: number) {
    const response = await apiClient.delete<ApiResponse<void>>(`/api/routings/${id}`)
    return response.data
  },

  async updateRoutingStatus(id: number, request: FactoryRoutingStatusUpdateRequest) {
    const response = await apiClient.patch<ApiResponse<FactoryRoutingResponse>>(`/api/routings/${id}/status`, request)
    return response.data
  },

  async getRoutingUsage(id: number) {
    const response = await apiClient.get<ApiResponse<FactoryRoutingUsageResponse>>(`/api/routings/${id}/usage`)
    return response.data
  },

  async getFactories() {
    const response = await apiClient.get<ApiResponse<string[]>>('/api/routings/factories')
    return response.data
  },

  async getLines(factoryName: string) {
    const response = await apiClient.get<ApiResponse<string[]>>(`/api/routings/factories/${encodePath(factoryName)}/lines`)
    return response.data
  },

  async getOperations(factoryName: string, lineName: string) {
    const response = await apiClient.get<ApiResponse<FactoryRoutingResponse[]>>(
      `/api/routings/factories/${encodePath(factoryName)}/lines/${encodePath(lineName)}/operations`,
      {
      }
    )
    return response.data
  },

  async getRoutingTree() {
    const response = await apiClient.get<ApiResponse<FactoryRoutingTreeResponse[]>>('/api/routings/tree')
    return response.data
  }
}

File: backend/src/main/java/com/ssafy/demo_app/api/routing/FactoryRoutingApi.java
package com.ssafy.demo_app.api.routing;

import com.ssafy.demo_app.api.routing.dto.FactoryRoutingRequest;
import com.ssafy.demo_app.api.routing.dto.FactoryRoutingResponse;
import com.ssafy.demo_app.api.routing.dto.FactoryRoutingStatusUpdateRequest;
import com.ssafy.demo_app.api.routing.dto.FactoryRoutingTreeResponse;
import com.ssafy.demo_app.api.routing.dto.FactoryRoutingUsageResponse;
import com.ssafy.demo_app.domain.routing.entity.FactoryRouting;
import com.ssafy.demo_app.global.response.ApiResponse;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;

import java.util.List;

@Tag(name = "Factory Routing API", description = "공장/라인/공정 라우팅 기준정보 관리 API")
@RequestMapping("/api/routings")
public interface FactoryRoutingApi {

    @Operation(summary = "라우팅 목록 조회", description = "공장명과 라인명으로 필터링하여 라우팅 목록을 조회합니다.")
    @GetMapping
    ResponseEntity<ApiResponse<List<FactoryRoutingResponse>>> getRoutings(
            @Parameter(description = "공장명 필터") @RequestParam(required = false) String factoryName,
            @Parameter(description = "라인명 필터") @RequestParam(required = false) String lineName,
            @Parameter(description = "라우팅 상태 필터") @RequestParam(required = false) FactoryRouting.RoutingStatus routingStatus
    );

    @Operation(summary = "라우팅 단건 조회", description = "ID로 특정 라우팅 정보를 조회합니다.")
    @GetMapping("/{routingId}")
    ResponseEntity<ApiResponse<FactoryRoutingResponse>> getRouting(
            @Parameter(description = "라우팅 ID", required = true) @PathVariable Integer routingId
    );

    @Operation(summary = "라우팅 등록", description = "공장명, 라인명, 공정 순서, 공정명을 등록합니다.")
    @PostMapping
    ResponseEntity<ApiResponse<FactoryRoutingResponse>> createRouting(
            @Valid @RequestBody FactoryRoutingRequest request
    );

    @Operation(summary = "라우팅 수정", description = "ID에 해당하는 라우팅 정보를 수정합니다.")
    @PutMapping("/{routingId}")
    ResponseEntity<ApiResponse<FactoryRoutingResponse>> updateRouting(
            @Parameter(description = "라우팅 ID", required = true) @PathVariable Integer routingId,
            @Valid @RequestBody FactoryRoutingRequest request
    );

    @Operation(summary = "라우팅 삭제", description = "작업지시에서 참조 중이지 않은 라우팅 정보를 삭제합니다.")
    @DeleteMapping("/{routingId}")
    ResponseEntity<ApiResponse<Void>> deleteRouting(
            @Parameter(description = "라우팅 ID", required = true) @PathVariable Integer routingId
    );

    @Operation(summary = "라우팅 상태 변경", description = "라우팅을 활성 또는 비활성 상태로 변경합니다.")
    @PatchMapping("/{routingId}/status")
    ResponseEntity<ApiResponse<FactoryRoutingResponse>> updateRoutingStatus(
            @Parameter(description = "라우팅 ID", required = true) @PathVariable Integer routingId,
            @Valid @RequestBody FactoryRoutingStatusUpdateRequest request
    );

    @Operation(summary = "라우팅 참조 현황 조회", description = "작업지시와 생산 실적에서 특정 라우팅을 참조하는 현황을 조회합니다.")
    @GetMapping("/{routingId}/usage")
    ResponseEntity<ApiResponse<FactoryRoutingUsageResponse>> getRoutingUsage(
            @Parameter(description = "라우팅 ID", required = true) @PathVariable Integer routingId
    );

    @Operation(summary = "공장 목록 조회", description = "라우팅에 등록된 공장명 목록을 조회합니다.")
    @GetMapping("/factories")
    ResponseEntity<ApiResponse<List<String>>> getFactories();

    @Operation(summary = "공장별 라인 목록 조회", description = "특정 공장에 등록된 라인명 목록을 조회합니다.")
    @GetMapping("/factories/{factoryName}/lines")
    ResponseEntity<ApiResponse<List<String>>> getLines(
            @Parameter(description = "공장명", required = true) @PathVariable String factoryName
    );

    @Operation(summary = "라인별 공정 목록 조회", description = "특정 공장/라인의 공정 목록을 순서대로 조회합니다.")
    @GetMapping("/factories/{factoryName}/lines/{lineName}/operations")
    ResponseEntity<ApiResponse<List<FactoryRoutingResponse>>> getOperations(
            @Parameter(description = "공장명", required = true) @PathVariable String factoryName,
            @Parameter(description = "라인명", required = true) @PathVariable String lineName
    );

    @Operation(summary = "라우팅 트리 조회", description = "공장 > 라인 > 공정 구조로 라우팅 정보를 조회합니다.")
    @GetMapping("/tree")
    ResponseEntity<ApiResponse<List<FactoryRoutingTreeResponse>>> getRoutingTree();
}
